### DISCUSSION TITLE
The Long Range wifi mode

### DISCUSSION THREAD

**Discussion by brokenbrigde9000** [micropython/micropython#11573] [ESP32]
I wanted to try the long range wifi mode in Micropython but this is less obvious than I thought. I have found a handful of (somewhat vague) references to this functionality in Micropython and at this point I'm not sure if it even exists. I try to send a message between 2 ESP32 devices (dfrobot firebeetle).

I found [this](https://github.com/micropython/micropython/issues/8510) thread referring to these lines:

w0 = network.WLAN(network.STA_IF)
w0.config(protocol=network.MODE_LR)

which sort of make sense to me but at the same time setting protocol=network.MODE_LR does not seem to do anything for me. When called before .active(True) the code sometimes runs once, but then throws me a 'RuntimeError: Wifi Unknown Error 0x0102' even after a hard reset. When called after .active(True) everything runs fine, but definitely not in the long range mode given the regular wifi performance.

I'm running [v1.14 (2021-02-02) .bin](http://micropython.org/resources/firmware/esp32-idf3-20210202-v1.14.bin) on both devices. Am I doing something wrong? Or does this functionality not exist yet?

Here is the code from my sending device:

```
import network
import espnow
import time

# A WLAN interface must be active to send()/recv()
sta = network.WLAN(network.STA_IF)  # Or network.AP_IF
sta.config(protocol=network.MODE_LR)
sta.active(True)

e = espnow.ESPNow()
e.active(True)
peer = b'\x10\x97\xbd\xdc\x12x'  # MAC address of peer's wifi interface
e.add_peer(peer)  # Must add_peer() before send()

e.send(peer, "Starting...")
for i in range(30):
    print('sending', i)
    e.send(peer, str(i) * 20, True)
    time.sleep(0.5)

e.send(peer, b'end')
print('finished transmission')
```

And here is the receiving end:

```
import network
import espnow

# A WLAN interface must be active to send()/recv()
sta = network.WLAN(network.STA_IF)
sta.config(protocol=network.MODE_LR)
sta.active(True)

e = espnow.ESPNow()
e.active(True)

print('start listening for data...')
while True:
    host, msg = e.recv()
    if msg:  # msg == None if timeout in recv()
        print(host, msg)
        peers = e.peers_table
        print('peer list', peers)


     if msg == b'end':
        break
print('finished receiving data')
```

**Comment by davefes** [comment]
A lot has happened to ESPNow since then.  I had a go at LR mode with v1.17 or 1.18 and couldn't see any difference.  Glenn has started a specific ESPNow thread.  There is also a proper wifi.py script that handles all the issues around getting the right sequences.
https://github.com/glenn20/micropython-espnow-utils/blob/main/src/wifi.py

**Reply by brokenbrigde9000** [reply]
Thanks for pointing out those tools to me. I appear to successfully get the devices in 'MODE_LR', and indeed broadcasting and receiving ESPNOW messages works. When calling wifi.status(), this is the result:

```
STA: on,  mac= 10:97:bd:dc:12:78 (b'\x10\x97\xbd\xdc\x12x')
AP : off, mac= 10:97:bd:dc:12:79 (b'\x10\x97\xbd\xdc\x12y')
     disconnected, channel=1, pm=1 (PM_PERFORMANCE), protocol=8 (MODE_LR)
```

However, the range of the connection is eye watering bad. Instead of a long range connection, I get one that reaches about half of a regular wifi connection. Am I still missing something? Maybe these ESP32 boards have really bad antennas? Their regular wifi performance is very good.

**Reply by davefes** [reply]
The antennas will perform the same for WiFi and ESPNow.  What is your range with ESPNow normal mode?

**Reply by brokenbrigde9000** [reply]
Regular wifi gets me about 25 meters with some obstacles (I leave the device on my desk and walk outside). When using LR-Mode the RSSI gets near -100 and no messages are received at a distance of <15 meters.

When connecting to my home network wifi router, I get about 35 meters with several obstacles (3 walls) in between when the ESP32 devices connect to it.

**Reply by davefes** [reply]
When you say "Regular WiFi" do you mean "Regular ESPNow", ie non-LR mode?   Sorry, I haven't tried LR mode recently.  Time to post on Glenn's ESPNow thread and see if he has done anything with LR-mode

**Reply by brokenbrigde9000** [reply]
Yes, with regular wifi I mean a non-LR mode (11B).

**Reply by davefes** [reply]
It was 1 year ago:
https://forum.micropython.org/viewtopic.php?f=5&t=12370&p=67124&hilit=ESPNow+LR#p67124
1-2dB increase and that could be due to environment changes as much as anything else.  Where are the pm constants documented?

**Reply by peterhinch** [reply]
Is measuring RSSI actually relevant? It's possible that RSSI is determined by transmitter power, receiver sensitivity, and antenna gain. These parameters may not change in LR mode. Yet it might implement a modulation mode enabling the receiver to more efficiently extract data from noise. Does anyone know how LR mode (and, for that matter RSSI measurement) claims to work?

**Reply by brokenbrigde9000** [reply]
I figured comparing RSSI between 11B mode and LR mode would make sense, but my knowledge doesn't go that far. Perhaps it does not tell anything about the performance and I should stick to the real-world test (actual distance).

I have more ESP32 units in my collection, including a couple of Espressif ESP32-WROOM-32E. I have flashed micropython on them and ran the same experiment. These devices defeat the DFRobot units by a factor 3 in distance. I can walk away up to 35 meters until the ESPNow messages stop coming in.

When connecting to my home network router, they all seem to have similar performance when it comes to maximum range. I guess we got MODE_LR working, but I'm a bit disappointed in the actual range. Perhaps an open field actually allows for a couple hundred meters, but add 2 trees and a wall and the LR mode seems to barely beat a normal wifi mode.

**Reply by davefes** [reply]
The bottom line on the quality of a radio link, in my experience, was determined by iPERF testing.  The amount of data you get through a radio channel at defined error-rates tells you quality.

Good summary on RSSI:
https://en.wikipedia.org/wiki/Received_signal_strength_indicator

Good point Peter.  I would expect that reducing receiver bandwidth or whatever processing changes they make should allow you to get a defined QoS at a lower RSSI.   Testing where a system falls-over is a rough indication of quality.

**Reply by brokenbrigde9000** [reply]
> The bottom line on the quality of a radio link, in my experience, was determined by iPERF testing. The amount of data you get through a radio channel at defined error-rates tells you quality.

Thanks for the links. In my usecase, getting _any_ data is pretty much good enough. If a sensor keeps blasting data is caught only once every 15 minutes at the 'gateway node', that would be fine.

I'm looking for several hundres of meters in range so I will probably resort to using Lora.

**Comment by davefes** [comment]
With common-variety WiFi whips at both ends, using ESPNow I typically get -73 to -76dBm at about 100 metres.  Worst case -78dm, that is still a good 15dB of margin.  I use LoRa for the longer links.

**Comment by glenn20** [comment]
I agree with @peterhinch here. I've had a lot of reports that LR_MODE doesn't make any difference to the RSSI values (but I wouldn't expect it to do so). The few reports that I have seen that demonstrate benefits of LR_MODE (not micropython examples) have shown a measured reduction in packet loss over distance. I think that will be the only convincing evidence for the benefits of LR_MODE.

Also - note that LR_MODE is not restricted to ESPNow usage. Espressif claims it can be used to improve long-range performance between any ESP32 devices. However, successful reports of the benefits are pretty few and far between.
